### ISSUE TITLE
Do not value copy structs when encountering struct pointers

### ISSUE THREAD

**Issue by deanm** [antonmedv/expr#53]
https://github.com/antonmedv/expr/blob/33a808e9f8835ba017273759a1441fb256b22ae4/vm/runtime.go#L44

Interfaces store a pair of either (type, ptr) or (type, value), the current runtime code will convert (type, ptr) to (type, value) and then recurse, but that conversion means creating a value copy, meaning value := *ptr.  This copy is definitely unnecessary and a performance concern for large/deep structures.

The solution is quite simple, the fetch function can unpack the ptr without going through an additional interface/function call, ex:

```
func fetch(from interface{}, i interface{}) interface{} {
	v := reflect.ValueOf(from)
	kind := v.Kind()

	if kind == reflect.Ptr {
		v = v.Elem()
		kind = v.Kind()
	}

	switch kind {

	case reflect.Array, reflect.Slice, reflect.String:
		index := toInt(i)
		value := v.Index(int(index))
		if value.IsValid() && value.CanInterface() {
			return value.Interface()
		}

	case reflect.Map:
		value := v.MapIndex(reflect.ValueOf(i))
		if value.IsValid() && value.CanInterface() {
			return value.Interface()
		}

	case reflect.Struct:
		value := v.FieldByName(reflect.ValueOf(i).String())
		if value.IsValid() && value.CanInterface() {
			return value.Interface()
		}
	}

	return nil
}
```

Additionally, you can support chasing through **T like:

```

	for ; kind == reflect.Ptr; kind = v.Kind() {
		v = v.Elem()
	}
```

This will then follow any amount of indirection until you get a type, meaning the environment could be a pointer to a pointer to a struct.

I'm happy to create a pull request with the change.

Thanks

**Comment by antonmedv** [comment]
Awesome! Can you create PR for this? If you also can provide a benchmark, it will be really interesting to see the performance increase.

**Comment by deanm** [comment]
PR #54 

Only a ~10000x performance improvement on my contrived example : )

**Comment by deanm** [comment]
I actually realized this problem is a general pattern in the runtime code.  There is a difference between a reflect.Value and putting that value into an interface, a reflect.Value is a way of accessing the pre-existing memory, where putting something into an interface is going to make a copy.

So another example in the code, if you have an embedded struct:

```
type Foo struct {
  Bar struct {
    Field int
  }
}
```

When you do 'Bar.Field', fetch will return a *copy* of Bar, not a pointer to the original Bar.  So if we do something to my previous example a stick a big chunk of memory in the Bar structure, it's going to value copy all of Bar.  Now that can be pretty easily fixed in the case of above (when Run() was called with an environment as a pointer), because then Bar's value will be addressable (CanAddr() is true) and you could create an interface of (type, *Bar) instead of (type, Bar)  by returning value.Addr() instead of value.  But I didn't look how it would work for other types, I guess slices are fine since the slice itself can be value copied (and what the slice points to isn't copy), but for fixed sized arrays, etc, I imagine the code now will value copy the entire array for an index operation like 'a[0]'.  The additional problem is CanAddr() will be false if you passed a value to Run() instead of a pointer, but at least this is something callers can easily control.

Basically I feel like it should be a principle of expr (for a few important reasons) to be zero-copy of the environment (when a caller calls Run() with a value and not a pointer, they create a copy at their callsite of course, but that doesn't really have anything to do with expr).  Right now it seems like expr is copying all over the place.

**Comment by deanm** [comment]
I tried making a patch for this but it's actually a bit harder than it seems, because I thought perhaps I could just return a *struct and a slice instead of an array in fetch(), but then the problem comes to method/functions calls when the arguments are pointers and they expected values, etc, and it seems pretty ugly to try to map that back at call sites, etc.  I think the real solution is probably working in reflect.Value's instead of interface{}.

**Comment by deanm** [comment]
See #56

**Comment by antonmedv** [comment]
> I think the real solution is probably working in reflect.Value's instead of interface{}.

I also was thinking about it.

**Comment by deanm** [comment]
#60

Easiest thing to do without a big change is just allow to opt in to the zero copy behaviour
